pkg/validation/validation.go

```diff
@@ -0,0 +1,220 @@
+		permanent[n] = true
+		sorted = append(sorted, n)
+		return nil
+	}
+	var keys []string
+	for k := range steps {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
+		if permanent[k] {
+			continue
+		}
+		if err := visit(k); err != nil {
+			return nil, err
+		}
+		if isCyclic {
+			return nil, k8sField.Forbidden(fieldPath, fmt.Sprintf("detected cycle %s", cycle))
+		}
+	}
+	return sorted, nil
+}
+
+func ValidateWorkflowSteps(steps map[string]api.WorkflowStep, fieldPath *k8sField.Path) k8sField.ErrorList {
+	allErrs := k8sField.ErrorList{}
+	if _, err := topologicalSort(steps, fieldPath); err != nil {
+		allErrs = append(allErrs, err)
+	}
+	for k, v := range steps {
+		if v.JobTemplate != nil && v.ExternalRef != nil {
+			allErrs = append(allErrs, k8sField.Invalid(fieldPath, k, "jobTemplate and externalRef are mutually exclusive"))
+		}
+	}
+	return allErrs
+}
+
+func ValidateWorkflowStatus(status *api.WorkflowStatus, fieldPath *k8sField.Path) k8sField.ErrorList {
+	allErrs := k8sField.ErrorList{}
+	return allErrs
+}
+
+func getWorkflowUnmodifiableSteps(workflow *api.Workflow) (running, completed map[string]bool) {
+	running = make(map[string]bool)
+	completed = make(map[string]bool)
+	if workflow.Status.Statuses == nil {
+		return
+	}
+	for key := range workflow.Spec.Steps {
+		if step, found := workflow.Status.Statuses[key]; found {
+			if step.Complete {
+				completed[key] = true
+			} else {
+				running[key] = true
+			}
+		}
+	}
+	return
+}
+
+func ValidateWorkflowUpdate(workflow, oldWorkflow *api.Workflow) k8sField.ErrorList {
+	allErrs := k8sValidation.ValidateObjectMetaUpdate(&workflow.ObjectMeta, &oldWorkflow.ObjectMeta, k8sField.NewPath("metadata"))
+
+	runningSteps, completedSteps := getWorkflowUnmodifiableSteps(oldWorkflow)
+	allCompleted := true
+	for k := range oldWorkflow.Spec.Steps {
+		if !completedSteps[k] {
+			allCompleted = false
+			break
+		}
+	}
+	if allCompleted {
+		allErrs = append(allErrs, k8sField.Forbidden(k8sField.NewPath("workflow"), "cannot update completed workflow"))
+		return allErrs
+	}
+
+	allErrs = append(allErrs, ValidateWorkflowSpecUpdate(&workflow.Spec, &oldWorkflow.Spec, runningSteps, completedSteps, k8sField.NewPath("spec"))...)
+	return allErrs
+}
+
+func ValidateWorkflowUpdateStatus(workflow, oldWorkflow *api.Workflow) k8sField.ErrorList {
+	allErrs := k8sValidation.ValidateObjectMetaUpdate(&oldWorkflow.ObjectMeta, &workflow.ObjectMeta, k8sField.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateWorkflowStatusUpdate(workflow.Status, oldWorkflow.Status)...)
+	return allErrs
+}
+
+func ValidateWorkflowSpecUpdate(spec, oldSpec *api.WorkflowSpec, running, completed map[string]bool, fieldPath *k8sField.Path) k8sField.ErrorList {
+	allErrs := k8sField.ErrorList{}
+	allErrs = append(allErrs, ValidateWorkflowSpec(spec, fieldPath)...)
+	allErrs = append(allErrs, k8sValidation.ValidateImmutableField(spec.Selector, oldSpec.Selector, fieldPath.Child("selector"))...)
```

Comment:
This will probably cause problems, as we update the selector from our own sync loop.